ptxboa/api_calc.py

```diff
@@ -1,87 +1,61 @@
 # -*- coding: utf-8 -*-
 """Classes for main process chain calculation."""
 
+import logging
+
 import pandas as pd
 
+from ptxboa.api_data import DataHandler
+
+logger = logging.getLogger()
 
-def get_from_dict(dct, key, msg=None):
-    """Raise detailed KeyError."""
-    try:
-        return dct[key]
-    except KeyError:
-        msg = f"{msg}: " if msg else ""
-        raise KeyError(f"{msg}'{key}' not in {list(dct)}.")
 
+def annuity(rate: float, periods: int, value: float) -> float:
+    """Calculate annuity.
 
-def get_from_df(df, key, msg=None):
-    """Raise detailed KeyError."""
-    try:
-        return df.loc[key]
-    except KeyError:
-        msg = f"{msg}: " if msg else ""
-        raise KeyError(f"{msg}'{key}' not in {list(df.index)}.")
+    Parameters
+    ----------
+    rate: float
+        interest rate per period
+    periods: int
+        number of periods
+    value: float
+        present value of an ordinary annuity
 
+    Returns
+    -------
+    : float
+        value of each payment
 
-def pmt(r, n, v):
-    if r == 0:
-        return v / n
+    """
+    if rate == 0:
+        return value / periods
     else:
-        return v * r / (1 - (1 / (1 + r) ** n))
+        return value * rate / (1 - (1 / (1 + rate) ** periods))
 
 
 class PtxCalc:
-    def __init__(
-        self,
-        data_handler,
-    ):
+    def __init__(self, data_handler: DataHandler):
         self.data_handler = data_handler
 
     def calculate(
```

Comment:
This is a long function. Can `calculate` be refactored to smaller functions? might be worth it if we need to touch it in the future again, otherwise let's keep it this way.

Reply:
yes,I want to go back to a class based implementation now that it works